### PR TITLE
Mosip 22871 download registraion centers

### PR DESCRIPTION
**Pull request by ramadeviidupulapati**

### CODE REVIEW

**Comment by ramadeviidupulapati** [comment]
closing this PR.

File: resident/resident-service/src/main/java/io/mosip/resident/controller/ProxyMasterdataController.java [hunk 1]
```diff
@@ -370,6 +384,31 @@ public ResponseWrapper<?> getAllTemplateBylangCodeAndTemplateTypeCode(@PathVaria
...
+	public ResponseEntity<Object> downloadRegistrationCentersByHierarchyLevel(@PathVariable("langcode") String langCode,
+			@PathVariable("hierarchylevel") Short hierarchyLevel, @RequestParam("name") List<String> name)
+			throws ResidentServiceCheckedException,IOException,Exception {
+		logger.debug("ProxyMasterdataController::getRegistrationCentersByHierarchyLevel()::entry");
+		auditUtil.setAuditRequestDto(EventEnum.GET_REG_CENTERS_FOR_LOCATION_CODE);	
+		validator.validateOnlyLanguageCode(langCode);
+		byte[] pdfBytes =  proxyMasterdataService.downloadRegistrationCentersByHierarchyLevel(langCode,hierarchyLevel, name);
+		InputStreamResource resource = new InputStreamResource(new ByteArrayInputStream(pdfBytes));
+		audit.setAuditRequestDto(EventEnum.DOWNLOAD_REGISTRATION_CENTER_SUCCESS);
+		System.out.println("after get service history pdf success");
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/controller/ProxyMasterdataController.java]: Remove system.out and only use loggers.

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 2]
```diff
@@ -413,5 +458,88 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
...
+				.getAllTemplateBylangCodeAndTemplateTypeCode(langCode, REGISTRATION_CENTER_TEMPLATE_NAME);		
+		ResponseWrapper<?> regCentResponseWrapper = proxyMasterdataService.getRegistrationCentersByHierarchyLevel(langCode,hierarchyLevel, name);
+		Map<String,Object> regCentersMap = new LinkedHashMap<>();
+		List<RegistrationCenterDto >regCentersDtlsList=Collections.EMPTY_LIST;
+		int serialNumber = 1;
+		if (regCentResponseWrapper != null) {
+			RegistrationCenterResponseDto registrationCentersDtls = mapper.readValue(
+					mapper.writeValueAsString(regCentResponseWrapper.getResponse()), RegistrationCenterResponseDto.class);
+			regCentersDtlsList = registrationCentersDtls.getRegistrationCenters();
+			for (RegistrationCenterDto regCenterDto : regCentersDtlsList) {
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: How big will this data be? for loops are not the efficient method.

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 3]
```diff
@@ -413,5 +458,88 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
...
+		Map<String, Object> templateResponse = new LinkedHashMap<>((Map<String, Object>) proxyResponseWrapper.getResponse());
+		String fileText = (String) templateResponse.get("fileText");
+		InputStream downLoadRegCenterTemplate = new ByteArrayInputStream(fileText.getBytes(StandardCharsets.UTF_8));
+		InputStream downLoadRegCenterTemplateData = templateManager.merge(downLoadRegCenterTemplate, regCentersMap);
+		
+		StringWriter writer = new StringWriter();
+		IOUtils.copy(downLoadRegCenterTemplateData, writer, "UTF-8");
+		ByteArrayOutputStream pdfValue = (ByteArrayOutputStream) pdfGenerator.generate(writer.toString());
+		logger.debug("ResidentServiceImpl::residentServiceHistoryPDF()::exit");
+		return pdfValue.toByteArray();
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: This is dangerous. We can not be doing this. Please look at webflux or spring async. How many calls are expected? this many byte arrays will kill our system.

**Review by gsasikumar — CHANGES_REQUESTED** [review]

File: resident/resident-service/src/main/java/io/mosip/resident/controller/ProxyMasterdataController.java (hunk 1 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/controller/ProxyMasterdataController.java]: removed  system.out.println statements and used loggers

File: resident/resident-service/src/main/java/io/mosip/resident/controller/ProxyMasterdataController.java [hunk 4]
```diff
@@ -370,6 +385,30 @@ public ResponseWrapper<?> getAllTemplateBylangCodeAndTemplateTypeCode(@PathVaria
...
+			throws ResidentServiceCheckedException,IOException,Exception {
+		logger.debug("ProxyMasterdataController::getRegistrationCentersByHierarchyLevel()::entry");
+		auditUtil.setAuditRequestDto(EventEnum.GET_REG_CENTERS_FOR_LOCATION_CODE);	
+		validator.validateOnlyLanguageCode(langCode);
+		InputStream pdfInputStream =  proxyMasterdataService.downloadRegistrationCentersByHierarchyLevel(langCode,hierarchyLevel, name);
+		InputStreamResource resource = new InputStreamResource(pdfInputStream);
+		audit.setAuditRequestDto(EventEnum.DOWNLOAD_REGISTRATION_CENTER_SUCCESS);
+		logger.debug("AcknowledgementController::acknowledgement()::exit");
+		return ResponseEntity.ok().contentType(MediaType.parseMediaType("application/pdf"))
+				.header("Content-Disposition", "attachment; filename=\"" + "downloadRegistrationCentersByHierarchyLevel" + ".pdf\"")
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/controller/ProxyMasterdataController.java]: dont hard code the name derive a name from date.

Comment (gsasikumar) [reply resident/resident-service/src/main/java/io/mosip/resident/controller/ProxyMasterdataController.java]: can we make it as regcenter-<date&time>?

File: resident/resident-service/src/main/java/io/mosip/resident/dto/RegistrationCenterDto.java [hunk 5]
```diff
@@ -56,6 +56,12 @@ public class RegistrationCenterDto extends MasterSyncBaseDto {
 	private String lunchEndTime;
 
 	private Boolean isActive;
+	
+	private String fullAddress;
+	
+	private int serialNumber;
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/dto/RegistrationCenterDto.java:62]: registration center has serial number?

File: resident/resident-service/src/main/java/io/mosip/resident/dto/WorkingDaysDto.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+import lombok.Data;
+
+/**
+ * 
+ * @author M1063027 Rama Devi
+ *
+ */
+@Data
+public class WorkingDaysDto {
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/dto/WorkingDaysDto.java:11]: The name of the DTO is not matching the actual content? is it working days?

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 7]
```diff
@@ -1,25 +1,50 @@
 package io.mosip.resident.service.impl;
 
+import java.io.ByteArrayInputStream;
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: remove all unwanted imports

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 8]
```diff
@@ -1,25 +1,50 @@
...
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.InputStream;
+import java.io.PipedInputStream;
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: do you need these many inputstreams?

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 9]
```diff
@@ -42,7 +71,28 @@ public class ProxyMasterdataServiceImpl implements ProxyMasterdataService {
...
+	@Autowired
+	private ProxyMasterdataService proxyMasterdataService;
+	
+	private TemplateManager templateManager;
 
+	@Autowired
+	private TemplateManagerBuilder templateManagerBuilder;
+
+	@PostConstruct
+	public void idTemplateManagerPostConstruct() {
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: why are we mixing variables and methods in a single place. All variables should be in the top and all member functions should be in the end.

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 10]
```diff
@@ -413,5 +462,118 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
 		logger.debug("ProxyMasterdataServiceImpl::getGenderTypesByLangCode()::exit");
 		return responseWrapper;
 	}
+  
+	/**
+	 * download registration centers based on language code, hierarchyLevel and center names
+	 */
+	public InputStream downloadRegistrationCentersByHierarchyLevel(String langCode, Short hierarchyLevel,
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: This method should not be in the proxy. It should be in a independent service.

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 11]
```diff
@@ -413,5 +462,118 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
...
+				.getAllTemplateBylangCodeAndTemplateTypeCode(langCode, REGISTRATION_CENTER_TEMPLATE_NAME);		
+		ResponseWrapper<?> regCentResponseWrapper = proxyMasterdataService.getRegistrationCentersByHierarchyLevel(langCode,hierarchyLevel, name);
+		Map<String,Object> regCentersMap = new LinkedHashMap<>();
+		List<RegistrationCenterDto >regCentersDtlsList=Collections.emptyList();		
+		if (regCentResponseWrapper != null) {
+			RegistrationCenterResponseDto registrationCentersDtls = mapper.readValue(
+					mapper.writeValueAsString(regCentResponseWrapper.getResponse()), RegistrationCenterResponseDto.class);
+			List<RegistrationCenterDto> regCenterIntialList = registrationCentersDtls.getRegistrationCenters();
+			if(regCenterIntialList!=null && !regCenterIntialList.isEmpty()) {
+				IntStream.range(0, regCenterIntialList.size()).forEach(i-> {
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: wow this is a killer. In some of the countries the list is so huge and going over one by one will be a disaster. why do we need to download the entire list? Is there are requirement to download all the center or just a specific hierarchy?

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 12]
```diff
@@ -413,5 +462,118 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
...
+	private void addRegistrationCenterDtls(int index, RegistrationCenterDto regCenterDto)
+			throws ResidentServiceCheckedException, Exception {
+		String workingHours = "";
+		String fullAddress = getFullAddress(regCenterDto.getAddressLine1(), regCenterDto.getAddressLine2(),
+				regCenterDto.getAddressLine3());
+		regCenterDto.setSerialNumber(index+1);
+		regCenterDto.setFullAddress(fullAddress);
+		List<WorkingDaysDto>  workingDaysList = getRegCenterWorkingDays(regCenterDto.getId(), regCenterDto.getLangCode());
+		workingHours = workingDaysList.get(0).getName() + "-" + workingDaysList.get(1).getName() + "|"
+				+ regCenterDto.getCenterStartTime().substring(0, 5) + " " + "am" + "-"
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: What if the center opens only at 12 noon? We should not be hard coding am or pm

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 13]
```diff
@@ -413,5 +462,118 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
...
+				+ regCenterDto.getCenterEndTime().substring(0, 5) + " " + "pm";
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: same as earlier do not hard code am or pm it should be derived.

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 14]
```diff
@@ -413,5 +462,118 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
...
+	 * @return
+	 */
+	private static InputStream convertOutputStreamToInputStream(ByteArrayOutputStream orgByteOutStream) {
+		PipedInputStream in = new PipedInputStream();
+		new Thread(new Runnable() {
+			public void run() {
+				try (final PipedOutputStream out = new PipedOutputStream(in)) {
+					orgByteOutStream.writeTo(out);
+				} catch (IOException e) {
+					logger.debug("convert Output stream to input stream" +e.getMessage());
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: it supposed to be logger.error. why is it debug?

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java [hunk 15]
```diff
@@ -413,5 +462,118 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
...
+	 * @param langCode
+	 * @return
+	 * @throws ResidentServiceCheckedException
+	 * @throws Exception
+	 */
+	private List<WorkingDaysDto> getRegCenterWorkingDays(String regCenterId, String langCode) throws ResidentServiceCheckedException,Exception {
+		ResponseWrapper<?> responseWrapper;
+		responseWrapper = proxyMasterdataService.getRegistrationCenterWorkingDays(regCenterId, langCode);
+		WorkingDaysResponseDto workingDaysResponeDtls = mapper.readValue(mapper.writeValueAsString(responseWrapper.getResponse()), WorkingDaysResponseDto.class);		
+		List<WorkingDaysDto> workingDaysList=workingDaysResponeDtls.getWorkingdays();
```
Comment (gsasikumar) [inline resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: How big is this data? Will it not be too big?

**Review by gsasikumar — CHANGES_REQUESTED** [review]

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 11 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: just a specific hierarchy. Based on user search criteria

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 12 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: Modified the code as per the review comments. View code changes in DownLoadMasterDataServiceImpl.java

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 13 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: Modified the code as per the review comments. View code changes in DownLoadMasterDataServiceImpl.java

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 15 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: very less data. Only get 6 working days of Data

File: resident/resident-service/src/main/java/io/mosip/resident/dto/RegistrationCenterDto.java (hunk 5 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/dto/RegistrationCenterDto.java:62]: No. Registration centers are not having the serial number. Serial Number is required as part of data generation in PDF.

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 10 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: Define the new controller and Service classes

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 3 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: modified the code as per the review comments and removed byte arrays

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 2 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: only get few registration center details data based on user search criteria

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 7 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: removed all unwanted imports

File: resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java (hunk 9 above)
Comment (ramadeviidupulapati) [reply resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java]: modified the code as per the review comments

**Review by loganathan-sekaran — APPROVED** [review]

**Review by gsasikumar — APPROVED** [review]

**Pull request merged**